provider/core/src/request.rs

```diff
@@ -8,111 +8,120 @@ use core::fmt;
 use core::fmt::Debug;
 use icu_locid::extensions::unicode as unicode_ext;
 use icu_locid::ordering::SubtagOrderingResult;
-use icu_locid::subtags::{Language, Region, Script, Variants};
 use icu_locid::{LanguageIdentifier, Locale};
 use writeable::{LengthHint, Writeable};
 
 #[cfg(doc)]
 use icu_locid::subtags::Variant;
 
-#[derive(Default, Debug, Clone, PartialEq, Eq)]
+#[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
 #[allow(clippy::exhaustive_structs)] // this type is stable
-pub struct DataRequest {
-    pub locale: DataLocale,
+pub struct DataRequest<'a> {
+    pub locale: DataLocale<'a>,
     pub metadata: DataRequestMetadata,
 }
 
-impl fmt::Display for DataRequest {
+impl fmt::Display for DataRequest<'_> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         fmt::Display::fmt(&self.locale, f)
     }
 }
 
-impl AsMut<DataLocale> for DataRequest {
-    fn as_mut(&mut self) -> &mut DataLocale {
-        &mut self.locale
-    }
-}
-
-#[derive(Default, Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+#[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
 #[non_exhaustive]
 pub struct DataRequestMetadata;
 
-/// A variant and language identifier, used for requesting data from a data provider.
-///
-/// The fields in a [`DataLocale`] are not generally known until runtime.
-#[derive(PartialEq, Clone, Default, Eq, Hash)]
-pub struct DataLocale {
-    langid: LanguageIdentifier,
-    keywords: unicode_ext::Keywords,
+#[derive(Clone, Copy, Debug)]
+#[allow(clippy::exhaustive_enums)] // private variants
+pub enum DataLocale<'a> {
+    Langid(&'a LanguageIdentifier),
+    Locale(&'a Locale),
 }
 
-impl fmt::Debug for DataLocale {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "DataLocale{{{}}}", self)
+impl PartialEq for DataLocale<'_> {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (DataLocale::Langid(s), DataLocale::Langid(o)) => s == o,
+            (DataLocale::Langid(s), DataLocale::Locale(o)) => {
+                *s == &o.id && o.extensions.unicode.keywords.is_empty()
+            }
+            (DataLocale::Locale(s), DataLocale::Langid(o)) => {
+                (&s.id) == *o && s.extensions.unicode.keywords.is_empty()
+            }
+            (DataLocale::Locale(s), DataLocale::Locale(o)) => {
+                s.id == o.id && s.extensions.unicode.keywords == o.extensions.unicode.keywords
```

Comment:
`question`: are we comfortable using `PartialEq` for two DataLocales that have different, non unicode, extensions? That seems like a nuanced gotcha.

Suggestion - forgo convenience and make it a method `cmp_relevant`? @Manishearth @sffc

Reply:
Alternative - if we have to keep it (because of Hash?) I'd suggest to document this gotcha in the struct top level description well.

Reply:
But the *`DataLocales`* don't have different non-unicode extensions. They don't have any. There's no way to observe a difference even if they're backed by `Locale`s with addtional extensions, so they are equal.